### ISSUE TITLE
assigning_clones: `clone_from` suggested on uninitialized variable

### ISSUE THREAD

**Issue by wanda-phi** [rust-lang/rust-clippy#12741]
### Summary

The `clippy::assigning_clones` diagnostic suggests calling `clone_from` on an uninitialized variable.

### Lint Name

clippy::assigning_clones

### Reproducer

I tried this code:

```rust
pub fn test() -> (String, String) {
    let (p, q): (String, String);
    p = "ghi".to_string();
    q = p.clone();
    (p, q)
}
```

I saw this happen:

```
warning: assigning the result of `Clone::clone()` may be inefficient
 --> src/lib.rs:4:5
  |
4 |     q = p.clone();
  |     ^^^^^^^^^^^^^ help: use `clone_from()`: `q.clone_from(&p)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assigning_clones
  = note: `#[warn(clippy::assigning_clones)]` on by default
```

Applying the suggestion of course causes an error, as `q` is not initialized at this point.

I expected to see this happen:

(no warning)

### Version

```text
rustc 1.79.0-beta.1 (6b544f5ff 2024-04-28)
binary: rustc
commit-hash: 6b544f5ff8d45221d61962651a5f5ab9fe535e16
commit-date: 2024-04-28
host: x86_64-unknown-linux-gnu
release: 1.79.0-beta.1
LLVM version: 18.1.4
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
